core/src/main/java/com/linecorp/armeria/server/metric/MetricCollectingService.java

```diff
@@ -88,11 +90,30 @@ public static MetricCollectingServiceBuilder builder(MeterIdPrefixFunction meter
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        if (ctx.config().transientServiceOptions().contains(TransientServiceOption.WITH_METRIC_COLLECTION)) {
+        if (shouldRecordMetrics(ctx)) {
             RequestMetricSupport.setup(ctx, REQUEST_METRICS_SET, meterIdPrefixFunction, true,
                                        successFunction != null ? successFunction::test
                                                                : ctx.config().successFunction());
         }
         return unwrap().serve(ctx, req);
     }
+
+    private boolean shouldRecordMetrics(ServiceRequestContext ctx) {
```

Comment:
I'm a bit worried that this might affect the performance in the worst case but this shouldn't be a problem in most cases.
We might add a cache later if it becomes a problem.

Reply:
Most applications, even complex business logic, may use less than 20~30 decorators.
I'm not sure how efficient the cache strategy is for them.

Do you want some numbers for the performance, I will try to run some benchmarks.

Reply:
I added a cache layer to check before looking for decorators with `.as()`. The size of `Route` is fixed when a server is built and it never changes until it is reconfigured.
All requests will hit the cache. Cache operations are not free though.

Reply:
Yeah, I thought it was not a problem in most cases so I didn't ask you the numbers because I didn't want to bother you. 😅 
Anyway, because you've already done some testing, could you share the numbers?

Reply:
> it never changes until it is reconfigured.

How about clearing the map in the `serviceAdded` method?

Reply:
It was added already at L138. 😉

Reply:
>  because you've already done some testing, could you share the numbers?

Oops, I misread your comments. 😓 Please ignore this. Thanks!